### PR TITLE
Add provider auto-registration

### PR DESCRIPTION
**Pull request by jianghaolu**
# ADD THIS TO DOCUMENT DB AND MONITOR AFTER THOSE 2 PRS ARE MERGED

### CODE REVIEW

**Comment by martinsawicki** [comment]
we'll need to remember to tweak the pending Monitor PR (manager) with this as well. Maybe apply directly to the PR, so we don't forget later?

**Comment by jianghaolu** [comment]
@martinsawicki Also the document db. I'll block this thread until we merge those 2.

**Comment by martinsawicki** [comment]
about MOnitor, i don't know if we're going to merge that for the next release. The reason we kept it from 1.0.0 (even as Beta) was because we had issues with the REST API exposing a nameless way of creating children (so our defineFoo() would have no params). We started a discussion about that with the service team.
Even if we identify a subset of Monitor we could merge, we do not currently have anyone to work on it. So not sure about that one...

OK to wait for DocDB at least.

File: azure-mgmt-resources/src/main/java/com/microsoft/azure/management/resources/fluentcore/utils/ProviderRegistrationInterceptor.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    public Response intercept(Chain chain) throws IOException {
+        Response response = chain.proceed(chain.request());
+        if (!response.isSuccessful()) {
+            String content = errorBody(response.body());
+            RestClient restClient = new RestClient.Builder()
+                    .withBaseUrl("https://" + chain.request().url().host())
+                    .withCredentials(credentials)
+                    .withSerializerAdapter(new AzureJacksonAdapter())
+                    .withResponseBuilderFactory(new AzureResponseBuilder.Factory())
+                    .build();
```
Comment (anuchandy) [inline azure-mgmt-resources/src/main/java/com/microsoft/azure/management/resources/fluentcore/utils/ProviderRegistrationInterceptor.java]: Given that `RestClient.Builder()...build()` is a heavy action, should we create RestClient only if `code` is `MissingSubscriptionRegistration`? so here we left with creation of `AzureJacksonAdapter`, that single instance can be used for both deserializing CloudError and with `withSerializerAdapter` wither?

File: azure-mgmt-resources/src/main/java/com/microsoft/azure/management/resources/fluentcore/utils/ProviderRegistrationInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+                Pattern pattern = Pattern.compile("/subscriptions/([\\w-]+)/");
+                Matcher matcher = pattern.matcher(chain.request().url().toString());
+                matcher.find();
+                ResourceManager resourceManager = ResourceManager.authenticate(restClient)
+                        .withSubscription(matcher.group(1));
+                pattern = Pattern.compile(".*'(.*)'");
+                matcher = pattern.matcher(cloudError.message());
+                matcher.find();
+                Provider provider = registerProvider(matcher.group(1), resourceManager);
+                while (provider.registrationState().equals("Unregistered")
```
Comment (anuchandy) [inline azure-mgmt-resources/src/main/java/com/microsoft/azure/management/resources/fluentcore/utils/ProviderRegistrationInterceptor.java]: `equalsIgnoreCase("Unregistered")`?

File: azure-mgmt-resources/src/main/java/com/microsoft/azure/management/resources/fluentcore/utils/ProviderRegistrationInterceptor.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+import java.io.IOException;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * An interceptor for automatic provider registration in Azure.
+ */
+public class ProviderRegistrationInterceptor implements Interceptor {
+    private AzureTokenCredentials credentials;
```
Comment (anuchandy) [inline azure-mgmt-resources/src/main/java/com/microsoft/azure/management/resources/fluentcore/utils/ProviderRegistrationInterceptor.java]: `private final`

File: azure-mgmt-resources/src/main/java/com/microsoft/azure/management/resources/fluentcore/utils/ProviderRegistrationInterceptor.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    private String errorBody(ResponseBody responseBody) throws IOException {
+        if (responseBody == null) {
+            return null;
+        }
+        BufferedSource source = responseBody.source();
+        source.request(Long.MAX_VALUE); // Buffer the entire body.
+        Buffer buffer = source.buffer();
+
```
Comment (anuchandy) [inline azure-mgmt-resources/src/main/java/com/microsoft/azure/management/resources/fluentcore/utils/ProviderRegistrationInterceptor.java]: empty line

File: azure-mgmt-resources/src/main/java/com/microsoft/azure/management/resources/fluentcore/utils/ProviderRegistrationInterceptor.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+            String content = errorBody(response.body());
+            RestClient restClient = new RestClient.Builder()
+                    .withBaseUrl("https://" + chain.request().url().host())
+                    .withCredentials(credentials)
+                    .withSerializerAdapter(new AzureJacksonAdapter())
+                    .withResponseBuilderFactory(new AzureResponseBuilder.Factory())
+                    .build();
+            CloudError cloudError = restClient.serializerAdapter().deserialize(content, CloudError.class);
+            if ("MissingSubscriptionRegistration".equals(cloudError.code())) {
+                Pattern pattern = Pattern.compile("/subscriptions/([\\w-]+)/");
```
Comment (anuchandy) [inline azure-mgmt-resources/src/main/java/com/microsoft/azure/management/resources/fluentcore/utils/ProviderRegistrationInterceptor.java]: do we need to make matching case-insensitive? 
```java
Pattern mypattern = Pattern.compile("/subscriptions/([\\w-]+)/", Pattern.CASE_INSENSITIVE);
```

**Comment by jianghaolu** [comment]
Addressed. Thanks!

**Comment by martinsawicki** [comment]
@anuchandy Hey Anu, are you able to sign off at this point?

**Comment by martinsawicki** [comment]
@jianghaolu DocDB mgmt was already merged - does anything have to be done to it as per the top remark?

**Comment by jianghaolu** [comment]
DocDB already updated. Merging.

**Pull request merged**